### PR TITLE
test(pipeline): add unit tests for execute_pipeline service #48

### LINKED ISSUE #48: Backend Unit Tests for Pipeline Executor
**Difficulty:** Hard
**Area:** backend, testing

## Description

The pipeline executor (`imagelab-backend/app/services/pipeline_executor.py`) is the central orchestration component that takes a list of pipeline steps, looks up each operator in the registry, and sequentially applies them to an image. Despite being the most critical backend service, it currently lacks unit test coverage.

Testing the executor is more complex than testing individual operators because it involves the registry lookup mechanism, sequential step execution, error propagation with step indices, and handling of edge cases like empty pipelines and unknown operator types. These tests are essential for ensuring pipeline reliability and providing clear error messages when something goes wrong.

## Acceptance Criteria

- [ ] Tests exist for the pipeline executor service
- [ ] Tests cover: empty pipeline, single-step pipeline, multi-step pipeline, unknown operator type, and error propagation with step index
- [ ] A reusable base64 image fixture is provided
- [ ] Tests use real operators (not mocks) for positive test cases
- [ ] Error test cases verify that the response contains meaningful error information
- [ ] All tests pass with `uv run pytest -v`
- [ ] Tests are independent and can run in any order

## Resources

- [pytest documentation](https://docs.pytest.org/en/stable/)
- Pipeline executor: `imagelab-backend/app/services/pipeline_executor.py`
- Pipeline models: `imagelab-backend/app/models/pipeline.py`
- Image utilities: `imagelab-backend/app/utils/image.py`
- Operator registry: `imagelab-backend/app/operators/registry.py`

### PR DESCRIPTION
**Pull request by Jayant-kernel**
## Description

Adds backend unit tests for the [execute_pipeline()](cci:1://file:///c:/Users/jayan/Desktop/c2si%20imag%20lab/imagelab-c2si-gsoc/imagelab-backend/app/services/pipeline_executor.py:7:0-44:5) function in [imagelab-backend/app/services/pipeline_executor.py](cci:7://file:///c:/Users/jayan/Desktop/c2si%20imag%20lab/imagelab-c2si-gsoc/imagelab-backend/app/services/pipeline_executor.py:0:0-0:0). Includes a shared pytest fixture for a reusable base64 image, and a `pythonpath` config fix so `uv run pytest -v` works out of the box.

Fixes #48

## Type of Change

- [ ] Bug fix
- [ ] New feature
- [ ] Breaking change
- [ ] Documentation update

## How Has This Been Tested?

Ran `cd imagelab-backend && uv run pytest -v` — all 7 tests pass.

- [ ] Existing tests pass
- [x] New tests added
- [ ] Manual testing

Tests cover: empty pipeline, NOOP step skipping, single real operator, multi-step chain, unknown operator with clear error, correct step index in errors, bad base64 decode, and test independence.

## Screenshots (if applicable)

N/A — backend only.

## Checklist

- [x] My code follows the project's style guidelines
- [x] I have performed a self-review
- [x] I have added/updated documentation as needed
- [x] My changes generate no new warnings
- [x] Tests pass locally

### CODE REVIEW

**Review by ivantha — COMMENTED** [review]
## ◎ Argus Review

This PR adds a well-structured unit test suite for the `execute_pipeline` service, including a clean `conftest.py` with a reusable factory fixture and a `pythonpath` fix for out-of-the-box `uv run pytest` support. Test coverage of error paths is notably thorough (unknown operator, step index, bad base64). The main concerns are weak assertions in success-path tests (no verification of actual transformations), a misleading test name for the independence test, and a semantic ambiguity around `step == 0` for decode errors. No critical bugs or security issues are present — this is a solid, low-risk addition.

File: imagelab-backend/tests/test_pipeline_executor.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+def test_error_includes_correct_step_index(make_request):
+    steps = [
+        PipelineStep(type="imageconvertions_grayimage"),
+        PipelineStep(type="bad_operator_step_one"),
+    ]
+    res = execute_pipeline(make_request(steps))
+    assert res.success is False
+    assert res.step == 1  # first step succeeds, second should fail
+    assert "bad_operator_step_one" in res.error
```
Comment (ivantha) [inline imagelab-backend/tests/test_pipeline_executor.py]: **[INFO]** Positive: Thorough error-path coverage with multi-dimensional assertions

`test_unknown_operator_gives_clear_error` and `test_error_includes_correct_step_index` both verify three orthogonal aspects of the failure response: `success is False`, the `step` index, and the content of the `error` message. This is a high-quality pattern for error testing — it guards against regressions in the error model (wrong step reported), the error message (unhelpful or wrong content), and the success flag simultaneously.

**Suggestion:**
```
No change needed. Extend this pattern to `test_error_includes_correct_step_index` (see finding about missing `"Unknown operator"` check) and `test_bad_image_data_fails_at_decode` for full consistency.
```
